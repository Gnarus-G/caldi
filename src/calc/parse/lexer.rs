#[derive(Clone, Copy, PartialEq)]
pub enum TokenKind {
    Ident,
    Integer,
    Plus,
    Eof,
    Illegal,
}

pub struct Token<'s> {
    pub kind: TokenKind,
    pub text: &'s str,
}

pub struct Lexer<'s> {
    input: &'s str,
    input_bytes: &'s [u8],
    position: usize,
    read_position: usize,
    tokens: Vec<Token<'s>>,
}

impl<'s> Lexer<'s> {
    pub fn new(input: &'s str) -> Self {
        Self {
            input,
            input_bytes: input.as_bytes(),
            position: 0,
            read_position: 0,
            tokens: vec![],
        }
    }

    pub fn tokenize(mut self) -> Vec<Token<'s>> {
        loop {
            if let Some(token) = self.tokens.last() {
                if token.kind == TokenKind::Eof {
                    break;
                }
            }

            self.next_token();
        }

        return self.tokens;
    }

    fn char(&self) -> Option<char> {
        self.input_bytes
            .get(self.position)
            .map(|&byte| byte as char)
    }

    fn peek_char(&self) -> Option<char> {
        self.input_bytes
            .get(self.read_position)
            .map(|&byte| byte as char)
    }

    fn advance(&mut self) {
        self.position = self.read_position;
        self.read_position += 1;
    }

    fn skip_whitespace(&mut self) {
        while self.char().unwrap_or('\0').is_ascii_whitespace() {
            self.advance();
        }
    }

    fn next_token(&mut self) {
        self.skip_whitespace();

        let c = match self.char() {
            Some(c) => c,
            None => {
                self.tokens.push(Token {
                    kind: TokenKind::Eof,
                    text: "",
                });
                return;
            }
        };

        let token = match c {
            '+' => Token {
                kind: TokenKind::Plus,
                text: "+",
            },

            c if c.is_ascii_digit() => {
                let start = self.position;

                while self.peek_char().unwrap_or('\0').is_ascii_digit() {
                    self.advance();
                }

                let end = self.position;

                let string = &self.input[start..=end];

                Token {
                    kind: TokenKind::Integer,
                    text: string,
                }
            }

            c if c.is_alphabetic() => {
                let start = self.position;

                while self.peek_char().unwrap_or('\0').is_alphabetic() {
                    self.advance();
                }

                let end = self.position;

                let string = &self.input[start..=end];

                let kind = match string {
                    "plus" => TokenKind::Plus,
                    _ => TokenKind::Ident,
                };

                Token { kind, text: string }
            }

            _ => Token {
                kind: TokenKind::Illegal,
                text: &self.input[self.position..self.position + 1],
            },
        };

        self.tokens.push(token);

        self.advance();
    }
}
