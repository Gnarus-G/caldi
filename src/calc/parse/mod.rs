use self::{
    ast::{BinOp, BinaryExpr, Expr, UnaryExpr},
    lexer::{Lexer, Token, TokenKind},
};

mod lexer;

pub struct Parser<'s> {
    tokens: Vec<Token<'s>>,
    position: usize,
    read_position: usize,
}

pub type Result<T> = std::result::Result<T, error::ErrorKind>;

impl<'s> Parser<'s> {
    pub fn new(input: &'s str) -> Self {
        Self {
            tokens: Lexer::new(input).tokenize(),
            position: 0,
            read_position: 1,
        }
    }

    fn token(&self) -> Option<&Token<'s>> {
        if let Some(t) = self.tokens.get(self.position) {
            if t.kind == TokenKind::Eof {
                return None;
            }

            return Some(t);
        }

        return None;
    }

    fn peek_token(&self) -> Option<&Token<'s>> {
        if let Some(t) = self.tokens.get(self.read_position) {
            if t.kind == TokenKind::Eof {
                return None;
            }

            return Some(t);
        }

        return None;
    }

    fn advance(&mut self) {
        self.position = self.read_position;
        self.read_position += 1;
    }

    pub fn parse(&mut self) -> Result<Expr> {
        self.parse_expr(Precedence::default())
    }

    fn parse_expr(&mut self, curr_precedence: Precedence) -> Result<Expr> {
        let mut exp = match self.token() {
            Some(token) => match token.kind {
                TokenKind::Ident => {
                    self.advance(); // skipping identifiers
                    self.parse()?
                }
                TokenKind::Integer => self.parse_number(),
                TokenKind::Plus => self.parse_unary_expr()?,
                TokenKind::Minus => self.parse_unary_expr()?,
                TokenKind::Times => {
                    return Err(error::ErrorKind::UnexpectedToken {
                        token: token.into(),
                    })
                }
                TokenKind::Over => {
                    return Err(error::ErrorKind::UnexpectedToken {
                        token: token.into(),
                    })
                }
                TokenKind::Eof => return Err(error::ErrorKind::UnexpectedEnd { at: token.start }),
                TokenKind::Illegal => {
                    self.advance(); // skipping any illegal characters
                    self.parse()?
                }
            },
            None => return Err(error::ErrorKind::UnexpectedEnd { at: 0 }),
        };

        let peek_precedence: Precedence = match self.peek_token().map(|t| t.try_into()) {
            Some(Ok(p)) => p,
            Some(Err(_)) => return Ok(exp),
            None => return Ok(exp),
        };

        while self
            .peek_token()
            .map(|t| t.kind != TokenKind::Eof)
            .unwrap_or(false)
            && curr_precedence < peek_precedence
        {
            self.advance();
            exp = self.parse_binary_expr(exp)?;
        }

        Ok(exp)
    }

    fn parse_number(&self) -> Expr {
        let token = self.token().unwrap();

        Expr::Interger(token.text.parse().expect(
            "failed to parse an ostensibly properly tokenized interger (should not happen)",
        ))
    }

    fn parse_unary_expr(&mut self) -> Result<Expr> {
        let op: ast::UnOp = match self.token().map(|t| t.try_into()) {
            Some(Ok(op)) => op,
            Some(Err(token)) => {
                return Err(error::ErrorKind::UnexpectedToken {
                    token: token.into(),
                })
            }
            None => return Err(error::ErrorKind::UnexpectedEnd { at: 0 }),
        };

        self.advance();

        let number = self.parse()?;
        Ok(Expr::UnExpr(Box::new(UnaryExpr { op, right: number })))
    }

    fn parse_binary_expr(&mut self, left: Expr) -> Result<Expr> {
        let op: BinOp = match self.token().map(|t| t.try_into()) {
            Some(Ok(op)) => op,
            Some(Err(token)) => {
                return Err(error::ErrorKind::UnexpectedToken {
                    token: token.into(),
                })
            }
            None => return Err(error::ErrorKind::UnexpectedEnd { at: 1 }),
        };

        self.advance();

        Ok(Expr::BinExpr(Box::new(BinaryExpr {
            left,
            op,
            right: self.parse_expr(op.into())?,
        })))
    }
}

#[derive(Debug, Default, PartialEq, PartialOrd)]
enum Precedence {
    #[default]
    None,
    Sum,
    Product,
}

impl From<BinOp> for Precedence {
    fn from(value: BinOp) -> Self {
        match value {
            BinOp::Plus => Self::Sum,
            BinOp::Minus => Self::Sum,
            BinOp::Times => Self::Product,
            BinOp::Over => Self::Product,
        }
    }
}

impl<'t, 's> TryFrom<&'t Token<'s>> for Precedence {
    type Error = &'t Token<'s>;

    fn try_from(value: &'t Token<'s>) -> std::prelude::v1::Result<Self, Self::Error> {
        BinOp::try_from(value).map(Precedence::from)
    }
}

pub mod ast {
    use super::lexer::Token;

    #[derive(Debug)]
    pub enum Expr {
        Interger(isize),
        BinExpr(Box<BinaryExpr>),
        UnExpr(Box<UnaryExpr>),
    }

    #[derive(Debug)]
    pub struct BinaryExpr {
        pub left: Expr,
        pub op: BinOp,
        pub right: Expr,
    }

    #[derive(Debug)]
    pub struct UnaryExpr {
        pub op: UnOp,
        pub right: Expr,
    }

    #[derive(Debug, Clone, Copy)]
    pub enum BinOp {
        Plus,
        Minus,
        Times,
        Over,
    }

    impl<'t, 's> TryFrom<&'t Token<'s>> for BinOp {
        type Error = &'t Token<'s>;

        fn try_from(value: &'t Token<'s>) -> Result<Self, Self::Error> {
            let r = match value.kind {
                super::lexer::TokenKind::Minus => BinOp::Minus,
                super::lexer::TokenKind::Times => BinOp::Times,
                super::lexer::TokenKind::Over => BinOp::Over,
                super::lexer::TokenKind::Plus => BinOp::Plus,
                _ => return Err(value),
            };

            Ok(r)
        }
    }

    #[derive(Debug)]
    pub enum UnOp {
        Plus,
        Minus,
    }

    impl<'t, 's> TryFrom<&'t Token<'s>> for UnOp {
        type Error = &'t Token<'s>;

        fn try_from(value: &'t Token<'s>) -> Result<Self, Self::Error> {
            let r = match value.kind {
                super::lexer::TokenKind::Minus => UnOp::Minus,
                super::lexer::TokenKind::Plus => UnOp::Plus,
                _ => return Err(value),
            };

            Ok(r)
        }
    }
}

pub mod error {
    use std::fmt::Display;

    use super::lexer::{Token, TokenKind};

    #[derive(Debug)]
    pub struct TokenKindAt {
        pub position: usize,
        pub kind: TokenKind,
    }

    impl From<&Token<'_>> for TokenKindAt {
        fn from(value: &Token<'_>) -> Self {
            Self {
                position: value.start,
                kind: value.kind,
            }
        }
    }

    #[derive(Debug)]
    pub enum ErrorKind {
        UnexpectedToken { token: TokenKindAt },
        UnexpectedEnd { at: usize },
    }

    impl std::error::Error for ErrorKind {}

    impl Display for ErrorKind {
        fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
            match self {
                ErrorKind::UnexpectedToken { token } => write!(
                    f,
                    "unexpected token {:?} at position {}",
                    token.kind, token.position
                ),
                ErrorKind::UnexpectedEnd { at } => write!(
                    f,
                    "unexpected end of expression encountered at position {}",
                    at
                ),
            }
        }
    }
}
