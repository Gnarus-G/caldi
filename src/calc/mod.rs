use std::{
    fmt::Display,
    ops::{Add, Div, Mul, Neg, Sub},
};

use self::parse::{
    ast::{BinaryExpr, Expr, UnaryExpr},
    Parser,
};

mod parse;

#[derive(Debug, PartialEq)]
enum Value {
    Integer(isize),
    Float(f64),
}

impl Add for Value {
    type Output = Value;

    fn add(self, rhs: Self) -> Self::Output {
        match (self, rhs) {
            (Value::Integer(l), Value::Integer(r)) => (l + r).into(),
            (Value::Integer(l), Value::Float(r)) => (l as f64 + r).into(),
            (Value::Float(l), Value::Integer(r)) => (l + r as f64).into(),
            (Value::Float(l), Value::Float(r)) => (l + r).into(),
        }
    }
}

impl Sub for Value {
    type Output = Value;

    fn sub(self, rhs: Self) -> Self::Output {
        match (self, rhs) {
            (Value::Integer(l), Value::Integer(r)) => (l - r).into(),
            (Value::Integer(l), Value::Float(r)) => (l as f64 - r).into(),
            (Value::Float(l), Value::Integer(r)) => (l - r as f64).into(),
            (Value::Float(l), Value::Float(r)) => (l - r).into(),
        }
    }
}

impl Mul for Value {
    type Output = Value;

    fn mul(self, rhs: Self) -> Self::Output {
        match (self, rhs) {
            (Value::Integer(l), Value::Integer(r)) => (l * r).into(),
            (Value::Integer(l), Value::Float(r)) => (l as f64 * r).into(),
            (Value::Float(l), Value::Integer(r)) => (l * r as f64).into(),
            (Value::Float(l), Value::Float(r)) => (l * r).into(),
        }
    }
}

impl Div for Value {
    type Output = Value;

    fn div(self, rhs: Self) -> Self::Output {
        match (self, rhs) {
            (Value::Integer(l), Value::Integer(r)) => (l as f64 / r as f64).into(),
            (Value::Integer(l), Value::Float(r)) => (l as f64 / r).into(),
            (Value::Float(l), Value::Integer(r)) => (l / r as f64).into(),
            (Value::Float(l), Value::Float(r)) => (l / r).into(),
        }
    }
}

impl Neg for Value {
    type Output = Value;

    fn neg(self) -> Self::Output {
        match self {
            Value::Integer(number) => (-number).into(),
            Value::Float(number) => (-number).into(),
        }
    }
}

impl Display for Value {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        match self {
            Value::Integer(number) => write!(f, "{number}"),
            Value::Float(number) => write!(f, "{number}"),
        }
    }
}

impl From<isize> for Value {
    fn from(value: isize) -> Self {
        Self::Integer(value)
    }
}

impl From<f64> for Value {
    fn from(value: f64) -> Self {
        Self::Float(value)
    }
}

pub fn eval(source: &str) -> parse::Result<String> {
    let mut parser = Parser::new(source);

    let expr = parser.parse()?;

    eprintln!("[DEBUG] ast: {expr:?}");

    Ok(eval_expr(&expr).to_string())
}

fn eval_expr(expr: &Expr) -> Value {
    match expr {
        Expr::Integer(i) => (*i).into(),
        Expr::BinExpr(expr) => eval_binary_expr(expr),
        Expr::UnExpr(expr) => eval_unary_expr(expr),
        Expr::Float(f) => (*f).into(),
    }
}

fn eval_binary_expr(expr: &BinaryExpr) -> Value {
    let left: Value = eval_expr(&expr.left);

    let right = eval_expr(&expr.right);

    match expr.op {
        parse::ast::BinOp::Plus => left + right,
        parse::ast::BinOp::Minus => left - right,
        parse::ast::BinOp::Times => left * right,
        parse::ast::BinOp::Over => left / right,
    }
}

fn eval_unary_expr(expr: &UnaryExpr) -> Value {
    let number = eval_expr(&expr.right);

    match expr.op {
        parse::ast::UnOp::Plus => number,
        parse::ast::UnOp::Minus => -number,
    }
}

pub fn render_error(error: parse::error::ErrorKind, source: &str) -> String {
    let mut output = String::new();

    let location = match &error {
        parse::error::ErrorKind::UnexpectedToken { token } => token.position,
        parse::error::ErrorKind::UnexpectedEnd { at } => *at,
    };

    output.push_str(source);

    output.push_str(&" ".repeat(location));
    output.push('\n');
    output.push_str(&" ".repeat(location));

    output.push('â†³');
    output.push(' ');

    output.push_str(&error.to_string());

    output
}

#[cfg(test)]
mod tests {
    use insta::assert_display_snapshot;

    use crate::calc::{eval, render_error};

    macro_rules! assert_evals {
        ($expr:literal, $ans:expr) => {
            assert_eq!(eval($expr).unwrap(), ($ans).to_string())
        };
    }

    #[test]
    fn ops_in_letters() {
        assert_evals!("minus 3", -3);
        assert_evals!("negative 3", -3);
        assert_evals!("2 plus 2 times 3 minus 1 over 2 minus negative 5", 12.5);
    }

    #[test]
    fn unaries() {
        assert_evals!("-2", -2);
        assert_evals!("--2", 2);
        assert_evals!("+2", 2);
        assert_evals!("-+-2", 2);
        assert_evals!("--+-2", -2);
    }

    #[test]
    fn sums() {
        assert_evals!("3 + 2", 5);
        assert_evals!("3 + 3.", 6);
        assert_evals!("3 - 2 / 6", 2.6666666666666665);
        assert_evals!("75 + 100", 175);
        assert_evals!("75 - 100", -25);
        assert_evals!("75 + 1000 - 100", 975);
        assert_evals!("1000 + 75 - 100", 975);
    }

    #[test]
    fn products() {
        assert_evals!("3 * 2", 6);
        assert_evals!("89 * 34", 3026);
        assert_evals!("89 * 34 * 23 * 199", 13850002);
        assert_evals!("9 / 2", 4.5);
        assert_evals!("256 * 9 / 2", 1152);
    }

    #[test]
    fn pemdas() {
        assert_evals!("9 * 2 / 3 + 6 - 4 + 2", 10);
        assert_evals!("2 + 2 / 2 - - 2", 5);
        assert_evals!("2 + 2 * 3 - 1 / 2 - - 5", 12.5);
        assert_evals!("3 - 1 / 2 - 5", -2.5);
        assert_evals!("-3 - 1 / 2 - 5", -8.5)
    }

    macro_rules! assert_error {
        ($source:literal) => {
            let source = $source;
            let err = eval(source).unwrap_err();
            let prettied = render_error(err, source);
            insta::with_settings!({ description => source }, {
                assert_display_snapshot!(prettied)
            })
        };
    }

    #[test]
    fn errors() {
        assert_error!("* 2");
        assert_error!("/ 2");
        assert_error!("2 + * 2");
    }
}
