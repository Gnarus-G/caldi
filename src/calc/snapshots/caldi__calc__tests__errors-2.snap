---
source: src/calc/mod.rs
description: / 2
expression: prettied
---
/ 2
â†³ unexpected token Over at position 0
